src/compiler/checker.ts

```diff
@@ -8506,13 +8537,15 @@ namespace ts {
                         return getUnionType(assignableConstituents);
                     }
                 }
-                // If the candidate type is assignable to the target type, narrow to the candidate type.
-                // Otherwise, if the current type is assignable to the candidate, keep the current type.
-                // Otherwise, the types are completely unrelated, so narrow to the empty type.
+                // If the candidate type is a subtype of the target type, narrow to the candidate type.
+                // Otherwise, narrow to whichever of the target type or the candidate type that is assignable

```

Comment:
I think there's a little more precision than the wording lends to here. If the candidate isn't a subtype, but is assignable, we keep the original type in place. (e.g. with `Array.isArray`, you preserve the original type instead of giving back `Array<any>`).